### PR TITLE
Support for image-based observations in GPI-LS and Envelope

### PR DESCRIPTION
**Pull request by LucasAlegre**
This PR should make it possible to run GPI-LS and Envelope on mo-supermario-v0 (https://mo-gymnasium.farama.org/environments/mo-supermario/) and other environments with image-based observations.

### CODE REVIEW

File: benchmark/launch_experiment.py [hunk 1]
```diff
@@ -189,11 +200,48 @@ def main():
         )
 
     else:
-        env = MORecordEpisodeStatistics(mo_gym.make(args.env_id), gamma=args.gamma)
-        eval_env = mo_gym.make(args.env_id)
+        if "mario" in args.env_id:
+            env = mo_gym.make(args.env_id, death_as_penalty=True)
```
Comment (ffelten) [inline benchmark/launch_experiment.py:204]: Can't we bundle this and time limit into the env registration?

File: benchmark/launch_experiment.py [hunk 2]
```diff
@@ -189,11 +200,48 @@ def main():
...
+                env = ResizeObservation(env, (84, 84))
+                env = GrayScaleObservation(env)
+                env = FrameStack(env, 4)
+                env = TimeLimit(env, max_episode_steps=5000)
+                return env
+
+            env = wrap_mario(env)
+            eval_env = wrap_mario(eval_env)
+
+            if args.record_video:
```
Comment (ffelten) [inline benchmark/launch_experiment.py:238]: Shouldn't be mario specific, I mean, we should be able to recordvideo of any env?

File: morl_baselines/common/utils.py [hunk 3]
```diff
@@ -63,3 +65,56 @@ def make_gif(env, agent, weight: np.ndarray, fullpath: str, fps: int = 50, lengt
     clip = ImageSequenceClip(list(frames), fps=fps)
     clip.write_gif(fullpath + ".gif", fps=fps)
     print("Saved gif at: " + fullpath + ".gif")
+
+
+class MaxAndSkipObservationV0(gym.Wrapper):
```
Comment (ffelten) [inline morl_baselines/common/utils.py:70]: This should be moved to MO-Gymnasium

File: morl_baselines/multi_policy/gpi_pd/gpi_pd.py [hunk 4]
```diff
@@ -619,12 +624,19 @@ def _reset_priorities(self, w: th.Tensor):
...
-        num_batches = int(np.ceil(obs_s.size(0) / 1000))
+        ) = self.replay_buffer.get_all_data(to_tensor=False)
+        num_batches = int(np.ceil(obs_s.shape[0] / 1000))
         for i in range(num_batches):
             b = i * 1000
-            e = min((i + 1) * 1000, obs_s.size(0))
+            e = min((i + 1) * 1000, obs_s.shape[0])
             obs, actions, rewards, next_obs, dones = obs_s[b:e], actions_s[b:e], rewards_s[b:e], next_obs_s[b:e], dones_s[b:e]
+            obs, actions, rewards, next_obs, dones = (
+                th.tensor(obs).to(self.device),
```
Comment (ffelten) [inline morl_baselines/multi_policy/gpi_pd/gpi_pd.py:634]: Why this change of not getting it as tensor from the buffer?

**Review by ffelten — CHANGES_REQUESTED** [review]
Thank you for this one, will be very useful for practical applications I think.
Could you upload/cleanup the runs on openrlbenchmarks (with video capture if possible?)

File: morl_baselines/common/utils.py (hunk 3 above)
Comment (LucasAlegre) [reply morl_baselines/common/utils.py:70]: I added to Gymnasium (https://github.com/Farama-Foundation/Gymnasium/releases/tag/v0.29.0). But then we need to require >=0.29.0

File: morl_baselines/multi_policy/gpi_pd/gpi_pd.py (hunk 4 above)
Comment (LucasAlegre) [reply morl_baselines/multi_policy/gpi_pd/gpi_pd.py:634]: I thought it would use less memory by not moving the entire replay buffer to GPU.

File: morl_baselines/common/utils.py (hunk 3 above)
Comment (ffelten) [reply morl_baselines/common/utils.py:70]: Awesome! I don't have anything against bumping the version in that case.

File: morl_baselines/multi_policy/gpi_pd/gpi_pd.py (hunk 4 above)
Comment (ffelten) [reply morl_baselines/multi_policy/gpi_pd/gpi_pd.py:634]: That's smart.

File: morl_baselines/common/utils.py (hunk 3 above)
Comment (LucasAlegre) [reply morl_baselines/common/utils.py]: Damn, the one in Gymnasium does not work with vector rewards. I will have to move this one to MO-Gymnasium.

File: benchmark/launch_experiment.py (hunk 1 above)
Comment (LucasAlegre) [reply benchmark/launch_experiment.py:204]: I think I prefer to leave this outside and let users explicitly configure these, as in Atari envs.
